vpr/src/place/analytic_placer.h

```diff
@@ -0,0 +1,249 @@
+ */
+template<typename T>
+struct EquationSystem;
+
+/*
+ * @brief helper function to find the index of macro that contains blk
+ * returns index in placementCtx.pl_macros, -1 if blk not in any macros
+ */
+int imacro(ClusterBlockId blk);
+
+/*
+ * @brief helper fucntion to find the head of macro containing blk
+ * returns the ID of the head block
+ */
+ClusterBlockId macro_head(ClusterBlockId blk);
+
+class AnalyticPlacer {
+  public:
+    /*
+     * @brief Constructor of AnalyticPlacer, currently initializes AnalyticPlacerCfg for the analytic placer
+     * To tune these parameters, change directly in constructor
+     */
+    AnalyticPlacer();
+
+    /*
+     * @brief top function of analytic placement
+     * Takes the random initial placement from place.cpp through g_vpr_ctx
+     * Repeat the following until stopping criteria is met:
+     * 	* Formulate and solve equations in x, y directions for 1 type of logial block
+     * 	* Instantiate CutSpreader to spread and strict_legalize
+     *
+     * The final legal placement is passed back to annealer in g_vpr_ctx.mutable_placement()
+     */
+    void ap_place();
+
+  private:
+    // for CutSpreader to access legal_pos, block_locs, etc.
+    friend class CutSpreader;
+
+    // AP parameters that can influence it's behavior
+    struct AnalyticPlacerCfg {
+        float alpha;                        // anchoring strength of pseudo-connections
+        float beta;                         // over-utilization factor
+        int criticalityExponent;            // not currently used, @see build_equations()
+        int timingWeight;                   // not currently used, @see build_equations()
+        float solverTolerance;              // parameter of the solver
+        int spread_scale_x, spread_scale_y; // see CutSpreader::expand_regions()
+    };
+
+    AnalyticPlacerCfg tmpCfg; // TODO: PlacerHeapCfg should be externally configured & supplied
+
+    // @see initial_placement.cpp load_legal_locations()
+    std::vector<std::vector<int>> num_legal_pos;
+    std::vector<std::vector<std::vector<t_pl_loc>>> legal_pos;
+
+    // marker for blks not solved in current iteration
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // row number in the system of linear equations for each block
+    std::unordered_map<ClusterBlockId, int32_t> blk_row_num;
+
+    // Encapsulates 3 types of locations for each logic block
+    struct BlockLocation {
+        t_pl_loc loc;       // real, up-to-date location of the logic block
+        t_pl_loc legal_loc; // legalized location, used to create psudo connections
+        double rawx, rawy;  // raw location storing float result from matrix solver
+    };
+
+    // Lookup from blockID to block location
+    std::unordered_map<ClusterBlockId, BlockLocation> blk_locs;
+
+    // The set of blks actually placed.
+    // Excludes children of macros.
+    std::vector<ClusterBlockId> place_blks;
```

Comment:
Explain what "children of macros" means (I think you mean non-head blocks of macros).  Does it also exclude blocks that are .fixed (locked down)?  I think you want to say "The ids of the blocks that are placed by the matrix solver, i.e. the free variable blocks. This excludes blocks that are part of placement macros but not the head of that macro, and fixed blocks."